### PR TITLE
kresd service: switch .listenDoH to new implementation

### PR DESCRIPTION
**Pull request by vcunat**
###### Motivation for this change
Switch the option to a better implementation.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- N/A Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- N/A Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- N/A Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
- - -
I don't use DoH normally, but I briefly tested that it looks OK.

There's this `extraFeatures` caveat (though very minor IMO), so I wanted to wait a while for feedback before merging.

### CODE REVIEW

**Comment by vcunat** [comment]
I considered also adding `.listenXDP`, but the technology seems useful only for relatively special performance-focused deployments, so I suspect such easy option wouldn't be useful anyway.

File: pkgs/servers/dns/knot-resolver/default.nix [hunk 1]
```diff
@@ -38,6 +39,11 @@ unwrapped = stdenv.mkDerivation rec {
     # ExecStart can't be overwritten in overrides.
     # We need that to use wrapped executable and correct config file.
     sed '/^ExecStart=/d' -i systemd/kresd@.service.in
+  ''
+    # some tests have issues with network sandboxing, apparently
+  + optionalString doInstallCheck ''
+    sed '1i os.exit(77)' -i daemon/lua/trust_anchors.test/bootstrap.test.lua
```
Comment (Mic92) [inline pkgs/servers/dns/knot-resolver/default.nix:45]: Maybe?

```
echo 'os.exit(77)' > daemon/lua/trust_anchors.test/bootstrap.test.lua
```

Comment (vcunat) [reply pkgs/servers/dns/knot-resolver/default.nix:45]: Well, why not 🤷🏽   Similar enough, same end result.  It doesn't seem important to me at all.

**Review by Mic92 — APPROVED** [review]
LGTM in general. Have not tested.

File: pkgs/servers/dns/knot-resolver/default.nix (hunk 1 above)
Comment (Mic92) [reply pkgs/servers/dns/knot-resolver/default.nix:45]: It is just slightly more readable.

**Comment by Mic92** [comment]
There is only one caveat when upgrading. It's no longer possible to put kresd/http2 behind an nginx reverse proxy. nginx only supports http1.1 to talk to the backend. Proxies like nginx are required if port 443 is already in use otherwise i.e. for normal https traffic. It is possible to use nginx stream module to match on the hostname but than it gets awkward to support normal https vhosts.

**Comment by Mic92** [comment]
In my deployment I can work around it by using sslh to match on the SNI hostname. However how do get new letsencrypt certificates in this case? There is no acme module available for kresd.

**Comment by Mic92** [comment]
I am using the web management backend for the time being: https://github.com/Mic92/dotfiles/commit/d5164e9c138f842a5658d8f64dbb9a189c17e62c

**Comment by vcunat** [comment]
Thanks for the feedback; I'll post more details later.

Detail: `kind = 'doh'` didn't go away in kresd 5.2.0, so it's possible to keep using that (i.e. it's possible to just revert this PR - or avoid using `.listenDoH` and configure it in `.extraConfig`).

**Comment by vcunat** [comment]
If you're interested in auto-reloading the certificate (which should be a fairly reliable methond, I expect), there's [a way](https://gitlab.nic.cz/knot/knot-resolver/-/issues/598#note_167796).  Note that it uses lua cqueues, so you'd still need it to add e.g. through `extraFeatures = true` (though that option's a little over-kill for that case).

Upstream docs for explicit reloads will be incoming, but I'd personally think that's a little harder to integrate.

**Comment by vcunat** [comment]
Honestly, the non-support in nginx is a surprise.  I understand their arguments, but... fortunately other proxies seem to usually support /2.

**Comment by Mic92** [comment]
> If you're interested in auto-reloading the certificate (which should be a fairly reliable methon, I expect), there's [a way](https://gitlab.nic.cz/knot/knot-resolver/-/issues/598#note_167796). Note that it uses lua cqueues, so you'd still need it to add e.g. through `extraFeatures = true` (though that option's a little over-kill for that case).
> 
> Better upstream docs for explicit reloads will be incoming soon, but I'd personally think that's a little harder to integrate.

This solves reload at least. Is it possible to serve the acme validation directory somehow? Than all the common ACME tools work again.

**Comment by vcunat** [comment]
No, `doh2` can only do DoH, no other HTTP stuff.  It's possible to use `doh2` for DoH and additionally `http` module on some other IP-port combinations (to serve whatever... probably not hard).

**Comment by Mic92** [comment]
> No, `doh2` can only do DoH, no other HTTP stuff. It's possible to use `doh2` for DoH and additionally `http` module on some other IP-port combinations (to serve whatever... probably not hard).

Different port won't work. Doh2 needs to be on port 443.

**Comment by vcunat** [comment]
I meant that the `http` for ACME might be on a different IP, for example.

**Comment by Mic92** [comment]
I don't see how multiple IPs would solve the issue. I think acme checks multiple ip addresses. However a simple hack that could be implemented into doh2 is to redirect to a different hostname: i.e. `dns.example.com/.well-known/acme-challenge/random-filename` -> `acme.dns.example.com/.well-known/acme-challenge/random-filename`. Than it could be handled by a different ip address.

**Comment by vcunat** [comment]
Right, you want to serve DoH on the IPs that are advertised in DNS.

Here's the docs around explicit reloads: https://gitlab.nic.cz/knot/knot-resolver/-/merge_requests/1098

**Comment by vcunat** [comment]
I'm no good around ACME.  By any chance, it's not possible to make it work just with port 80, right?

**Comment by Mic92** [comment]
> I'm no good around ACME. By any chance, it's not possible to make it work just with port 80, right?

Oh, you are right. Why did I not thought about this!

**Comment by vcunat** [comment]
For reference, my reading is that the HTTP challenge only uses port 80:

> This request MUST be sent to TCP port 80 on the HTTP server.

https://tools.ietf.org/id/draft-ietf-acme-acme-18.html#rfc.section.8.3

In particular, "doh2" in kresd does not even implement insecure http, so it can't be used to affect this ACME challenge.  (The answer over port 80 may be a redirect to https, but in that case you can surely redirect to a better place instead.)

**Pull request merged**